contracts/token/psp22/src/extensions/wrapper.rs

```diff
@@ -0,0 +1,84 @@
+/// Extension of [`PSP22`] which supports token wrapping
+use crate::traits::*;
+
+use brush::{
+    declare_storage_trait,
+    traits::{
+        AccountId,
+        Balance,
+        InkStorage,
+    },
+};
+use ink_prelude::vec::Vec;
+use ink_storage::traits::SpreadLayout;
+pub use psp22_derive::PSP22WrapperStorage;
+
+#[cfg(feature = "std")]
+use ink_storage::traits::StorageLayout;
+
+#[derive(Default, Debug, SpreadLayout)]
+#[cfg_attr(feature = "std", derive(StorageLayout))]
+pub struct PSP22WrapperData {
+    pub underlying: AccountId,
+}
+
+declare_storage_trait!(PSP22WrapperStorage, PSP22WrapperData);
+
+#[brush::wrapper]
+pub type PSP22WrapperWrapper = dyn PSP22WrapperExt + PSP22;
+
+#[brush::trait_definition]
+pub trait PSP22WrapperExt: PSP22WrapperStorage + PSP22 {
+    /// Allow a user to deposit `amount` of underlying tokens and mint `amount` of the wrapped tokens to `account`
+    #[ink(message)]
+    fn deposit_for(&mut self, account: AccountId, amount: Balance) -> Result<(), PSP22Error> {
+        self.deposit(amount)?;
+        self._mint(account, amount)
+    }
+
+    /// Allow a user to burn `amount` of wrapped tokens and withdraw the corresponding number of underlying tokens to `account`
+    #[ink(message)]
+    fn withdraw_to(&mut self, account: AccountId, amount: Balance) -> Result<(), PSP22Error> {
+        self._burn(Self::env().caller(), amount)?;
+        self.withdraw(account, amount)
+    }
+
+    /// Mint wrapped token to cover any underlyingTokens that would have been transfered by mistake. Internal
+    /// function that can be exposed with access control if desired.
+    fn _recover(&mut self, account: AccountId) -> Result<Balance, PSP22Error> {
+        let value = self.underlying_balance() - self.total_supply();
+        self._mint(account, value)?;
+        Ok(value)
+    }
+
+    /// helper function to transfer the underlying token from caller to the contract
+    fn deposit(&mut self, amount: Balance) -> Result<(), PSP22Error> {
+        let account_id = PSP22WrapperStorage::get_mut(self).underlying;
+        PSP22Wrapper::transfer_from(
+            &account_id,
+            Self::env().caller(),
+            Self::env().account_id(),
+            amount,
+            Vec::<u8>::new(),
+        )
+    }
+
+    /// helper function to transfer the underlying token
+    fn withdraw(&mut self, account: AccountId, amount: Balance) -> Result<(), PSP22Error> {
+        let account_id = PSP22WrapperStorage::get_mut(self).underlying;
```

Comment:
Instead of several private method to do a call will be easy to create a getter for `underlying`
```rust
fn _underlying(&mut self) -> &mut PSP22Wrapper {
    &mut PSP22WrapperStorage::get_mut(self).underlying
}